### PR TITLE
Merge .NET Core support into the portable libraries

### LINKED ISSUE #575: Add support for ASP.NET 5 and the new Core CLR
I tried to create an NUnit test project using Visual Studio 2015 using the ASP.NET 5 class library that uses the Core CLR. It built fine against the full CLR, but failed to compile against the Core CLR.

Web development is probably one of the areas that .NET is most widely used in, so we should be ready for the next version of ASP.NET.

**Update**

At this point, the new portable NUnit framework built using the `nunit-dnx.sln` targets the following frameworks;

- .NET Framework 4.6 (net46)
- ASP.NET Core 5.0 (dnxcore50) and
- Windows Universal 10.0 (uwp.10.0)

This is intended to allow testing of cross-platform ASP.NET 5 applications, Windows 10 Universal apps and .NET Core applications.

For information on installing and using ASP.NET 5 on Windows, see https://docs.asp.net/en/latest/index.html

The nunitlite.runner assembly will allow you to create a new .NET Core Console Application that will self run any unit tests contained within itself.

image

**Outstanding Tasks**

- Add the new assemblies to the framework and nunitlite NuGet packages
- Add the ability to build, test and package the new projects to NUnit.proj
- Add a project template to nunit/nunit.templates
- Create a more full featured test runner including running tests from Visual Studio

### PR DESCRIPTION
**Pull request by rprouse**
Fixes #575 and #876 

Based on his work with XUnit, @bradwilson recommended in #575 that we switch to using portable profile 259 which will allow the portable version of the framework to be used in the following platforms,

- .NET 4.5+
- .NET Core (Universal Windows Apps 10+, DNX Core 5+)
- Windows 8
- Windows Phone 8 (Silverlight)
- Windows Universal (Windows Phone 8.1+, Windows 8.1+)
- Xamarin (MonoTouch, MonoAndroid, Xamarin iOS Universal)
- Portable Libraries (supporting Profile259)

I have removed the existing DNX solution and projects and merged the NETCORE changes into the existing portable framework. This is a large pull request, but the majority is just merging the various `PORTABLE` and `NETCORE` `#defines`

This also resulted in the removal of SilverLight from the portable framework, thus the fix for #876.

Many of the reflection API's that we rely on are not available in this profile. To keep as much of our existing code from changing, I created extension methods for everything that we use that is missing and added an extensive test suite to ensure they behave the same as their counterparts in the desktop framework.

Portable now has an NUnitLite package. It doesn't use Console, but you can create a console runner from it by passing Console.Out/In into AutoRun. On platforms without a console, you can pass in a TextWriter/TextReader. The portable assembly was added to the NUnitLite nuget package.

The portable version of NUnitLite drops support for a color console, but it could be re-added in the future if we switched NUnitLite to use an interface that wraps the console.

I also added the NUnitLite.Tests to the portable project. NUnit.proj has been updated to build and run all portable projects and their tests.

@CharliePoole, could you pull this branch and see if you can build it? I think it requires updating to VS 2015. The changes also need testing on CF. I am currently on a machine without VS2008.

I still want to do some testing with the NuGet packages and different platforms, but this is ready to merge. Any additional problems can be fixed in subsequent issues.

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs [hunk 1]
```diff
@@ -399,9 +399,7 @@ public CultureInfo CurrentCulture
             set
             {
                 _currentCulture = value;
-#if NETCORE
-                CultureInfo.CurrentCulture = _currentCulture;
-#elif !NETCF
+#if !NETCF && !PORTABLE
```
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: We cannot set the CurrentCulture in this profile. If anyone knows another way...

File: src/NUnitFramework/framework/Compatibility/ReflectionExtensions.cs [hunk 2]
```diff
@@ -59,7 +59,379 @@ public static Type GetTypeInfo(this Type type)
...
-#elif NETCORE
+    }
+
+#elif PORTABLE
+
+    /// <summary>
+    /// Provides NUnit specific extensions to aid in Reflection
+    /// across multiple frameworks
+    /// </summary>
+    public static class TypeExtensions
```
Comment (rprouse) [inline src/NUnitFramework/framework/Compatibility/ReflectionExtensions.cs]: This is the meat of this PR. Tests for this are below. The majority of the other changes are just merging the PORTABLE and NETCORE #defines.

File: src/NUnitFramework/nunitlite.runner/TextRunner.cs [hunk 3]
```diff
@@ -214,24 +216,24 @@ public void ReportResults(ITestResult result)
         {
             Summary = new ResultSummary(result);
 
-            _textUI.DisplaySummaryReport(Summary);
+            if (Summary.SkipCount + Summary.IgnoreCount > 0)
```
Comment (rprouse) [inline src/NUnitFramework/nunitlite.runner/TextRunner.cs]: I re-ordered the NUnitLite output to match the recent changes to the console.

**Comment by CharliePoole** [comment]
@rprouse I'll pull and test either tonight or in the morning. I'll have to do some installing first.

File: src/NUnitFramework/testdata/nunit.testdata-4.0.csproj [hunk 4]
```diff
@@ -41,28 +41,34 @@
...
-    <Reference Include="System.Data" />
-    <Reference Include="System.Net" />
-    <Reference Include="System.Xml" />
-    <Reference Include="Microsoft.Threading.Tasks, Version=1.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+    <Reference Include="Microsoft.Threading.Tasks">
       <HintPath>..\..\..\packages\Microsoft.Bcl.Async.1.0.168\lib\net40\Microsoft.Threading.Tasks.dll</HintPath>
+      <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Threading.Tasks.Extensions, Version=1.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+    <Reference Include="Microsoft.Threading.Tasks.Extensions">
```
Comment (rprouse) [inline src/NUnitFramework/testdata/nunit.testdata-4.0.csproj]: Uninstalling BCL.Async from the portable projects caused the other projects to update, thus these changes.

File: src/NUnitFramework/tests/Compatibility/ReflectionExtensionsTests.cs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+using NUnit.Framework.Compatibility;
+
+namespace NUnit.Framework.Tests.Compatibility
+{
+    /// <summary>
+    /// A series of unit tests to ensure that the type extensions in the portable
+    /// framework behave the same as their counterparts in the full framework
+    /// </summary>
+    [TestFixture]
+    public class ReflectionExtensionsTests
```
Comment (rprouse) [inline src/NUnitFramework/tests/Compatibility/ReflectionExtensionsTests.cs]: These are the unit tests for the new type extensions. They run in all frameworks to ensure that there is no difference between the built in types and the extensions.

File: src/NUnitFramework/nunitlite.runner/TextRunner.cs (hunk 3 above)
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite.runner/TextRunner.cs]: Cool! I've been meaning to do that.

**Comment by CharliePoole** [comment]
Code looks good and tests pass (I have not yet run CF) so I'll just make general comments.

1. Our standard so far has been to name the assemblies the same for all builds. You have at least one that has "portable" in the name. Is that by design?

2. We only have a portable nunitlite, because it has to reference the framework, and it references the portable framework. The framework tests are not portable, but reference both the framework and nunitlite. This would work equally well if the "portable" runner were not actually portable, but was a .NET 4.5 application. We might have an two nunitlites per platform supported - one that references the native framework and one that references the the portable framework. I'm not sure whether this would be needed for anything except testing NUnit, however. We may want to discuss this after merging.

**Comment by rprouse** [comment]
The portable in the name was intentional since there are no other equivalent projects. We can rename?

**Comment by CharliePoole** [comment]
I see. I just saw the name nunit.portable.tests.exe and thought it was some kind of tests. It's actually a single-line runner. If you can think of a name that won't confuse me, feel free to change. :-)

If not we can wait till we think of one.

**Comment by CharliePoole** [comment]
A number of your ReflectionExtensionTests fail in CF. I'll fix and push but probably not before tomorrow. 

Also, the "reversing" of the console display doesn't work well when viewing it in a scrollable box - that is in CF and Silverlight. You have to scroll down to see the result. We might want it to be in a different order in those platforms. This is minor and could wait for another PR though.

**Comment by rprouse** [comment]
It is interesting that the reflection tests fail on CF. Those test all pass on the desktop framework so to me that means that some of the reflection APIs behave differently on CF than they do on desktop, so we might end up having different behaviour on that platform. None of the extensions are used except on portable, so test failures are just differences in the frameworks :frowning:

**Comment by CharliePoole** [comment]
CF Reflection is big-time-different!

**Comment by rprouse** [comment]
Maybe just #ifdef out all those tests for CF? They are mainly intended to
ensure portable acts like desktop.

**Comment by CharliePoole** [comment]
If there are no references to the extensions, I could #ifdef them out as
well.

**Comment by rprouse** [comment]
The extensions are already #ifdef'd out for all platforms except portable
except for gettypeinfo which is a straight passthrough.

**Comment by CharliePoole** [comment]
I think this is ready to merge.

**Comment by rprouse** [comment]
Your additional changes look good to me.

**Comment by ctaggart** [comment]
This is really great to see coreclr support! I'm really looking forward to a new NuGet build with this so that we can get the Visual F# tests working for coreclr. https://github.com/Microsoft/visualfsharp/issues/697

**Comment by rprouse** [comment]
@ctaggart, if only we were pushing NuGet packages to MyGet.org from CI builds! One of too many things on our ToDo list and not enough hands :smile: 

I am testing the NuGet now with Xamarin and DNX. If everything looks good, I can get you a copy for testing.

**Comment by ctaggart** [comment]
You can have your AppVeyor builds create a NuGet feed. You could easily create:
https://ci.appveyor.com/nuget/nunit
I explain how here: http://blog.ctaggart.com/2015/07/open-source-net-with-appveyor-nuget.html

**Comment by ctaggart** [comment]
I'm working on setting this up. I may have have a pull request ready in about an hour.

**Pull request merged**